design/authenticators/authn-k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,195 @@
+
+We would like to inject the cert and get an indication that the cert was injected  or the error if it was not. Let's explore the different options that we have to inject the cert. 
+
+**Note**: The SPDY protocol [is being deprecated](https://github.com/kubernetes/enhancements/issues/384)  so it's not an option.
+
+#### Option 1 - Run a parallel request to verify cert injection
+In this option, we send the tar request as we do today, but also create a parallel WebSocket in which we will send another request to verify that the cert was successfully injected. We have several options for the verification command:
+1. Run an "ls" command on the cert path. This way we get an immediate feedback on the  cert presence. However, this may return a positive answer even if the cert is not fully injected (this assumption will be investigated as part of this effort).
+1. Calculate the md5 hash of the cert content before sending it and then sending  a request to get the calculated md5 hash of the cert in the container. Once the two hashes are equal then we know that the cert was fully installed.
+
+#### Option 2 - Send the cert content as part of the URL
+In this option, we do not use STDIN to send the cert but send it in the URL. This eliminates the bug in KubeCtl Exec as without STDIN we will get the response and would be able to know whether the injection succeeded or not, and log the outcome.
+
+However, we may have a limit on the size of the URL and we will have an issue  if the certificate will grow in the future. To overcome this, we can split the  request. The first one will create the file. The following requests will populate it with the cert content. Then the last request will verify that the whole content has passed (with one of the methods described above). We can decide on the number  of requests dynamically according to the size of the cert.
+
+#### Option 3 - Client sets its State once cert is injected
+This option is relevant only if the "inject-client-cert" request will return a 202 Accepted response instead of 200 OK (more on that in [this section](#inject-client-cert-response-code).
+
+In this case, we respond with 202 Accepted to the "inject-client-cert" request and start injecting the cert in a separate thread. The server then polls on the state of the authn-client container and waits until its State changes to some  agreed State (e.g "CertInjected"). This state will be set by the authn-client once it sees that the cert is installed.
+
+This option is not fully backwards-compatible and is quite complex so I wouldn't suggest to implement it.
+
```

Comment:
**Option 4 - Combination of strategies**
Obviously waiting to write completion ack is very time consuming. Such approach might severely overload the server and its workers. Alternatively, for the common case, we can offloading of the work to the client such that:
1. Like today, the client will send a request, get response and polling the cert up to X seconds, before timeout. The server will inject CERT without waiting.
2. If timeout, send again inject_client_cert but this time with ack_on_write indication. Only this time the server will wait on inject CERT completion/error.

Reply:
I think timeout is a good idea, but an asynchronous retry after it might bring the system again to the condition of missing cert file

Reply:
Ofira, I am not sure i understand your comment. On retry client shift its strategy and send the request again with ack_on_write indication to the server- which is what suggested on the other paper.

Reply:
In this case it is a good idea - it was not clear to me originally

Reply:
@moticless in your solution how do you prevent form the client to fetch a cert which is not fully injected ? this will be really hard to troubleshoot

Reply:
@moticless the client could get the hash from the server and verify it, so when i rethinking about it this is not a problem

Reply:
@moticless so basically you're saying that we will add a header to the `inject_client_cert` request (e.g `Wait-On-Injection`). If the header is present then we will wait until the injection is done and return 200 OK if succeeded. If the header is not present we will not wait and return 202 Accepted. 

I think that changing the response code upon a header presence is a bit troublesome, did i miss something?

also - i don't think we have enough value in this suggestion to add this complexity. with this suggestion i would just wait for the timeout every time.

Last point - i don't think that this is really a solution suggestion and fits more in the `Inject Client Cert response code` section. wdyt?

Reply:
This option also adds backwards-compatibility issues as an old server will not do anything with this header.